Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/D3D12Quad.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <stdexcept>
+#include <wincodec.h>
+
+using namespace DirectX;
+using Microsoft::WRL::ComPtr;
+
+class D3D12Quad
+{
+public:
+    D3D12Quad(UINT width, UINT height, std::wstring name);
+
+    void OnInit();
+    void OnUpdate();
+    void OnRender();
+    void OnDestroy();
+
+    // Accessors.
+    UINT GetWidth() const { return m_width; }
+    UINT GetHeight() const { return m_height; }
+    const WCHAR* GetTitle() const { return m_title.c_str(); }
+
+    ComPtr<ID3D12Resource> GetCurrentBuffer();
+    void ShowNextImage();
+
+    bool is_initialized = false;
+
+
+private:
+    static const UINT FrameCount = 2;
+
+    struct Vertex
+    {
+        XMFLOAT3 position;
+        XMFLOAT2 uv;
+    };
+
+    // Pipeline objects.
+    CD3DX12_VIEWPORT m_viewport;
+    CD3DX12_RECT m_scissorRect;
+    ComPtr<IDXGISwapChain3> m_swapChain;
+    ComPtr<ID3D12Device> m_device;
+    ComPtr<ID3D12Resource> m_renderTargets[FrameCount];
+    ComPtr<ID3D12CommandAllocator> m_commandAllocator;
+    ComPtr<ID3D12CommandQueue> m_commandQueue;
+    ComPtr<ID3D12RootSignature> m_rootSignature;
+    ComPtr<ID3D12DescriptorHeap> m_rtvHeap;
+    ComPtr<ID3D12DescriptorHeap> m_srvHeap;
+    ComPtr<ID3D12PipelineState> m_pipelineState;
+    ComPtr<ID3D12GraphicsCommandList> m_commandList;
+    UINT m_rtvDescriptorSize;
+
+    // App resources.
+    ComPtr<ID3D12Resource> m_vertexBuffer;
+    D3D12_VERTEX_BUFFER_VIEW m_vertexBufferView;
+
+    // Synchronization objects.
+    UINT m_frameIndex;
+    HANDLE m_fenceEvent;
+    ComPtr<ID3D12Fence> m_fence;
+    UINT64 m_fenceValue;
+
+    void LoadPipeline();
+    void LoadAssets();
+    void PopulateCommandList();
+    void WaitForPreviousFrame();
+    void CreateRootSignature();
+    void CreatePipelineState();
+    void CreateVertexBuffer();
+    void CreateFence();
+    void CreateDescriptorHeaps();
+    void CreateFrameResources();
+    void CreateCurrentBuffer();
+    void CopyTextureIntoCurrentBuffer();
+
+    ID3D12Resource* textureBuffer; // the resource heap containing our texture
```

Comment:
>textureBuffer [](http://example.com/codeflow?start=20&length=13)

I don't see this being initialized to nullptr here or in the constructor, nor do I see it being freed later. Leak? We should free it in the destructor (if not null, then Release), or better yet, just use WRL ComPtr. (no need to pull in all of WinRT imo, as WRL is much lighter and makes it easier to integrate into various projects for people who want to take these snippets out)

Reply:
Same with `mainDescriptorHeap` and `textureBufferUploadHeap`.

Reply:
Done.